src/main/java/ch/njol/skript/lang/VariableString.java

```diff
@@ -542,33 +573,49 @@ public String toString(@Nullable Event e, boolean debug) {
 		StringBuilder b = new StringBuilder("\"");
 		for (Object o : string) {
 			if (o instanceof Expression) {
-				b.append("%").append(((Expression<?>) o).toString(e, debug)).append("%");
+				b.append("%").append(((Expression<?>) o).toString(event, debug)).append("%");
 			} else {
 				b.append(o);
 			}
 		}
 		b.append('"');
 		return b.toString();
 	}
-	
-	public String getDefaultVariableName() {
+
+	/**
+	 * Builds all possible default variable type hints based on the super type of the expression.
+	 * 
+	 * @return List<String> of all possible super class code names.
+	 */
+	public List<String> getDefaultVariableNames(String variableName, Event event) {
 		if (isSimple) {
 			assert simple != null;
-			return simple;
+			return Lists.newArrayList(simple, "object");
 		}
 		Object[] string = this.string;
 		assert string != null;
-		StringBuilder b = new StringBuilder();
-		for (Object o : string) {
-			if (o instanceof Expression) {
-				b.append("<")
-					.append(Classes.getSuperClassInfo(((Expression<?>) o).getReturnType()).getCodeName())
-					.append(">");
-			} else {
-				b.append(o);
+		List<StringBuilder> typeHints = Lists.newArrayList(new StringBuilder());
+		DefaultVariables data = script.getData(DefaultVariables.class);
+		assert data != null : "default variables not present in current script";
+
+		// Represents the index of which expression in a variable string, example name::%entity%::%object% the index of 0 will be entity.
+		int hintIndex = 0;
+		for (Object object : string) {
+			if (!(object instanceof Expression)) {
+				typeHints.forEach(builder -> builder.append(object));
+				continue;
+			}
+			StringBuilder[] current = typeHints.toArray(new StringBuilder[0]);
+			for (ClassInfo<?> classInfo : Classes.getAllSuperClassInfos(data.get(variableName)[hintIndex])) {
+				for (StringBuilder builder : current) {
+					String hint = builder.toString() + "<" + classInfo.getCodeName() + ">";
+					typeHints.add(new StringBuilder(hint));
+					typeHints.remove(builder);
```

Comment:
Can't you just append to the builder?

Reply:
No, we're adding a new builder here. We want array `[test::1::<commandsender>, test::1::<object>]` for `{test::1::%command sender%}`
```diff
- String hint = builder.toString() + "<" + classInfo.getCodeName() + ">";
+ String hint = builder.append("<" + classInfo.getCodeName() + ">").toString();
```
and
```diff
- String hint = builder.toString() + "<" + classInfo.getCodeName() + ">";
- typeHints.add(new StringBuilder(hint));
- typeHints.remove(builder);
+ builder.append("<" + classInfo.getCodeName() + ">");
```
Results in array `[test::1::<commandsender>, test::1::<commandsender><object>]`